core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -177,15 +177,17 @@ private[spark] class HostLocalDirManager(
  * Manager running on every node (driver and executors) which provides interfaces for putting and
  * retrieving blocks both locally and remotely into various stores (memory, disk, and off-heap).
  *
- * Note that [[initialize()]] must be called before the BlockManager is usable.
+ * Note that [[initialize()]] must be called before the BlockManager is usable. Also, the
+ * `memoryManager` is initialized at a later stage after DriverPlugin is loaded, to allow the
+ * plugin to overwrite memory configurations.
  */
 private[spark] class BlockManager(
     val executorId: String,
     rpcEnv: RpcEnv,
     val master: BlockManagerMaster,
     val serializerManager: SerializerManager,
     val conf: SparkConf,
-    memoryManager: MemoryManager,
+    var memoryManager: MemoryManager,
```

Comment:
Do we want to follow the same pattern as what `shuffleManager` does here ?

Reply:
I tried that at the beginning, but found out in certain cases there may be race conditions in:

```scala
private lazy val _memoryManager = Option(_memoryManager).getOrElse(SparkEnv.get.memoryManager)
```

Since a different thread can call `SparkEnv.set` right after the `memoryManger` is updated in the current `SparkEnv`. As result, the `memoryManager` could be null. This is revealed in `JobCancellationSuite`. 

The current approach makes the `memoryManager` a mutable field and updated later when the driver plugin is loaded.

Reply:
I am not sure I follow the corner case - can you point me to which test is causing this issue ? Thanks !
`BlockManager` being used would be associated with the corresponding `SparkEnv` - and if `SparkEnv` is being mutated, the new env is what we should be referencing.

Reply:
Sure. It is "job group with interruption". I think it is flaky though and doesn't always happen. When I tried this locally it doesn't always reproduce. The job link: https://github.com/sunchao/spark/actions/runs/7923522243/job/21637267400

```
[info] - job group with interruption *** FAILED *** (34 milliseconds)
[info]   java.lang.NullPointerException: Cannot invoke "org.apache.spark.memory.MemoryManager.maxOnHeapStorageMemory()" because the return value of "org.apache.spark.storage.BlockManager.memoryManager()" is null
[info]   at org.apache.spark.storage.BlockManager.maxOnHeapMemory$lzycompute(BlockManager.scala:243)
[info]   at org.apache.spark.storage.BlockManager.maxOnHeapMemory(BlockManager.scala:243)
[info]   at org.apache.spark.storage.BlockManager.initialize(BlockManager.scala:565)
[info]   at org.apache.spark.SparkContext.<init>(SparkContext.scala:633)
[info]   at org.apache.spark.SparkContext.<init>(SparkContext.scala:159)
[info]   at org.apache.spark.SparkContext.<init>(SparkContext.scala:172)
[info]   at org.apache.spark.JobCancellationSuite.$anonfun$new$41(JobCancellationSuite.scala:397)
[info]   at org.scalatest.enablers.Timed$$anon$1.timeoutAfter(Timed.scala:127)
[info]   at org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:282)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:230)
[info]   at org.apache.spark.SparkFunSuite.failAfter(SparkFunSuite.scala:69)
```

Reply:
I think we should be able to change `shuffleManager` in the same manner to avoid the potential concurrency issue.

Reply:
I'm missing how this is failing, can you clarify?  The blockmanager initialize is called on line 632 after the memory manager is initialized in the block manager, so how do we get a null?

Reply:
My suspicion is that after:

```scala
 _env.initializeMemoryManager(SparkContext.numDriverCores(master, conf))
```

in `SparkContext`, `SparkEnv.set` is called by a different thread, which updates the current `SparkEnv` to be a new instance without a `_memoryManager` initialized. Later on when the `maxOnHeapMemory` is called in `BlockManager.initialize`, it calls:

```scala
private lazy val _memoryManager = Option(_memoryManager).getOrElse(SparkEnv.get.memoryManager)
```

which will call `SparkEnv.get.memoryManager`. Since the active `SparkEnv` instance is updated, the `memoryManager` now is null.

I don't know how this would happen though, since it seems `SparkEnv` should be a singleton within a single JVM. Somehow I wasn't not able to reproduce this locally.

Reply:
I was able to reproduce the failure locally after many tries.

So the root cause is that when a task is running, it needs to access the `memoryStore` in a block manager (see [here](86575ad0-8f6f-422f-b10c-ea920f95caf5)). The access of `memoryStore` in turn will call `memoryManager`, which could be null in the following scenario:

1. a previous test launched a bunch of tasks which may not finish when the next test is started
2. the next test started, and calls `new SparkContext` which initialize everything include updating the active `SparkEnv` via `SparkEnv.set`
3. one of the left over tasks in the step 1 accesses (in a different thread than 2) `memoryStore` during its `run` method, which will then call `SparkEnv.get.memoryManager`. However, the `SparkEnv.get` method will return the new `SparkEnv` instance created in step 2, which, at this point may not have the `memoryManager` initialized yet!

I'm not sure whether this is a valid case. Looking at the [`Task` class](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/Task.scala), it calls `SparkEnv.get` in several places. However, in the rare cases when a `SparkContext` has been shut down and a new `SparkContext` is created, running tasks launched by the former could access `SparkEnv` created by the latter, which seems not valid. I think this can only happen in the local mode.

My latest change doesn't fix the issue either. I guess I just got lucky in the unit tests.

One approach I think, is not to use `SparkEnv.get` in the `Task` class, but rather pass in the active environment used by the `Executor`. Let me try this approach.

Reply:
Updated `Task` to use the block manager from the `SparkEnv` from the `Executor` that creates the task, instead of the one from `SparkEnv.get`. Also switched back to the original approach similar to the shuffle manager PR. The tests look good now.

Reply:
If this is a task from some previous test, which has completed, the exception is beneign at best and not a concern for non-test codepaths, right ?

If yes, do we need to try to work around this ? (other than by fixing the offending tests perhaps ?)

Reply:
I think it may also happen in the non-test path but could be very rare. I described the scenario above:

> I'm not sure whether this is a valid case. Looking at the [Task class](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/Task.scala), it calls SparkEnv.get in several places. However, in the rare cases when a SparkContext has been shut down and a new SparkContext is created, running tasks launched by the former could access SparkEnv created by the latter, which seems not valid. I think this can only happen in the local mode.

The [latest changes](https://github.com/apache/spark/pull/45052/files/a6502e06ef2bf68654f1d3c8bb5c483a071d8770..af68ece323401d06f7e036c3b0782bc80a0f4c93) updated `Task` to use the `BlockManager` from the `Executor` that launches the task, instead of the one from `SparkEnv.get` which could be from a different `SparkContext`.

Reply:
Users dont set `SparkEnv` - only spark does. For tests, we set `SparkEnv` for specific internal scenarios for validation.

As I mentioned, if `SparkContext` was shutdown - it is reasonable for tasks from previous ctx to fail (they should have been terminated anyway).

Reply:
> it is reasonable for tasks from previous ctx to fail

Hmm, if this is reasonable, then we just need to fix the test case instead (specifically `JobCancellationSuite`) and wait all tasks to be finish after each test.

Reply:
Although I think the change on the `Task` also makes sense though.

Reply:
@sunchao , I was looking at this again.
The stack trace of the NPE that we saw earlier was part of spark context initialization ... not an access from task, right ?

Given this, I am not sure if the hypothesis is valid ... am I missing something here ?

Reply:
I am still looking, but I believe this might be due to a race in `reportHeartBeat` in `Executor`.
Would like more eyes on it though.

My hypothesis is, something like this is happening:

* scheduler is [started here](https://github.com/apache/spark/blob/af68ece323401d06f7e036c3b0782bc80a0f4c93/core/src/main/scala/org/apache/spark/SparkContext.scala#L611).
  * In local mode, this directly ends up creating an `Executor`
    * TaskSchedulerImpl.start -> LocalSchedulerBackend.start -> new LocalEndpoint -> new Executor
* new Executor immediately starts a `heartbeater` - and the initial delay for the thread to start is random, after which it will periodically run.
* `Executor.reportHeartBeat`, which is invoked, sends a `Heartbeat`.
* If this `Heartbeat` is received before driver thread reaches [blockmanager.initialize](https://github.com/apache/spark/blob/af68ece323401d06f7e036c3b0782bc80a0f4c93/core/src/main/scala/org/apache/spark/SparkContext.scala#L632), it will end up returning `HeartbeatResponse` with `reregisterBlockManager = true` (since registeration of local block manager happens within `initialize`).

And as part of this, we end up initializing the lazy val memoryManager in BlockManager.

What I am unsure of is, this would normally have triggered an NPE anyway during reregisteration if this path is hit (without this PR) - not sure if that is getting ignored (since blockmanager id is null).

(I did not find driver logs for the test run - that would have helped validate this hypothesis - we would have seen "re-registering with master" for this test, before the NPE is thrown)

Thoughts ?

Reply:
I will need to look more - I dont this my hypothesis is the RC here ... even though `_env.blockManager.initialize` happens after `_taskScheduler.start()`,  `_env.initializeMemoryManager` happens before and should have initialized the state.

Reply:
> The stack trace of the NPE that we saw earlier was part of spark context initialization ... not an access from task, right ?

Thanks @mridulm for checking! I think that stack trace doesn't reveal the root cause of the issue. I added a bunch of debugging messages in the code and found out the task that was causing the issue:

```
setting active env to org.apache.spark.SparkEnv@5ab3ee8b in pool-1-thread-1-ScalaTest-running-JobCancellationSuite
active env = org.apache.spark.SparkEnv@5ab3ee8b, thread = Executor task launch worker for task 0.0 in stage 0.0 (TID 0)
java.base/java.lang.Thread.getStackTrace(Thread.java:1619)
org.apache.spark.storage.BlockManager.memoryManager$lzycompute(BlockManager.scala:210)
org.apache.spark.storage.BlockManager.memoryManager(BlockManager.scala:204)
org.apache.spark.storage.BlockManager.memoryStore$lzycompute(BlockManager.scala:248)
org.apache.spark.storage.BlockManager.memoryStore(BlockManager.scala:247)
org.apache.spark.scheduler.Task.$anonfun$run$3(Task.scala:146)
org.apache.spark.util.Utils$.tryLogNonFatalError(Utils.scala:1287)
org.apache.spark.scheduler.Task.run(Task.scala:144)
org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:633)
org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:96)
org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:636)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
java.base/java.lang.Thread.run(Thread.java:840)
memory manager of org.apache.spark.SparkEnv@5ab3ee8b is null, _memoryManager = null, thread = Executor task launch worker for task 0.0 in stage 0.0 (TID 0)
set memory manager for org.apache.spark.SparkEnv@5ab3ee8b, threadName = pool-1-thread-1-ScalaTest-running-JobCancellationSuite
java.base/java.lang.Thread.getStackTrace(Thread.java:1619)
org.apache.spark.SparkContext.<init>(SparkContext.scala:584)
org.apache.spark.SparkContext.<init>(SparkContext.scala:141)
org.apache.spark.JobCancellationSuite.$anonfun$new$45(JobCancellationSuite.scala:430)
org.scalatest.enablers.Timed$$anon$1.timeoutAfter(Timed.scala:127)
org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:282)
org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)
org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:230)
org.apache.spark.SparkFunSuite.failAfter(SparkFunSuite.scala:69)
org.apache.spark.SparkFunSuite.$anonfun$test$2(SparkFunSuite.scala:155)
org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
org.scalatest.Transformer.apply(Transformer.scala:22)
org.scalatest.Transformer.apply(Transformer.scala:20)
org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
```

The "setting active env" and "set memory manager" messages are logged in `SparkContext` initialization, while the "active env =" and "memory manager of " are logged in `BlockManager` when trying to access the `memoryManager`. The first stack trace shows it is from the separate worker thread.

Reply:
Sorry, the line number may not match since I added several changes in my local repo for debugging purpose.

Reply:
Thanks @mridulm . I like your solution which is simpler. Saving the `blockManager` at the beginning of `Task.run` should be sufficient. Let me adapt the code and the test case in this PR.